scipy/signal/_short_time_fft.py

```diff
@@ -0,0 +1,1659 @@
+* MyPy version 1.1.1 does not seem to support ``NDArray[complex]`` (only
+  ``NDArray[np.complex128]``). Hence the methods `stft()`, `stft_detrend()`,
+  `istft()`, and  `_ifft_func()` need to use `NDArray` array as a parameter
+  instead of `NDArray[complex]`.
+* The entry ``.. currentmodule:: scipy.signal.ShortTimeFFT`` is required
+  in docstrings to ensure that `stft` and `istft` generate HTML links to
+  methods and not the legacy functions. The side effect is that `ShortTimeFFT`
+  does not work anymore. It is not known how to do this file-wide instead of
+  method-wide.
+* The HTML documentation currently renders each method/property on a separate
+  page without reference to the parent class. Thus, a link to `ShortTimeFFT`
+  was added to the "See Also" section of each method/property.
+"""
+# Linter does allow to import ``Generator`` from ``typing`` module:
+from collections.abc import Generator
+from functools import cache, lru_cache, partial
+from typing import Callable, get_args, Literal, Optional, Union
+
+import numpy as np
+from numpy.typing import NDArray
+
+import scipy.fft as fft_lib
+from scipy.signal import detrend
+from .windows import get_window
+
+__all__ = ['ShortTimeFFT']
+
+
+#: Allowed values for parameter `padding` of method `ShortTimeFFT.stft()`:
+PAD_TYPE = Literal['zeros', 'edge', 'even', 'odd']
+
+#: Allowed values for property `ShortTimeFFT.fft_typ`:
+FFT_TYP_TYPE = Literal['twosided', 'centered', 'onesided', 'onesided2X']
+
+
+def _calc_dual_canonical_window(win: NDArray, hop: int) -> NDArray:
+    """Calculate canonical dual window for 1d window `win` and a time step
+    of `hop` samples.
+
+    A ``ValueError`` is raised, if the inversion fails.
+
+    This is a separate function not a method, since it is also used in the
+    class method ``ShortTimeFFT.from_dual()``.
+    """
+    if hop > len(win):
+        raise ValueError(f"{hop=} is larger than window length of {len(win)}" +
+                         " => STFT not invertible!")
+    if issubclass(win.dtype.type, np.integer):
+        raise ValueError("Parameter 'win' cannot be of integer type, but " +
+                         f"{win.dtype=} => STFT not invertible!")
+        # The calculation of `relative_resolution` does not work for ints.
+        # Furthermore, `win / DD` casts the integers away, thus an implicit
+        # cast is avoided, which can always cause confusion when using 32-Bit
+        # floats.
+
+    w2 = win.real**2 + win.imag**2  # win*win.conj() does not ensure w2 is real
+    DD = w2.copy()
+    for k_ in range(hop, len(win), hop):
+        DD[k_:] += w2[:-k_]
+        DD[:-k_] += w2[k_:]
+
+    # check DD > 0:
+    relative_resolution = np.finfo(win.dtype).resolution * max(DD)
+    if not np.all(DD >= relative_resolution):
+        raise ValueError("Short-time Fourier Transform not invertible!")
+
+    return win / DD
+
+
+# noinspection PyShadowingNames
+class ShortTimeFFT:
```

Comment:
There is a long list of methods/attributes, could you add a list in the docstrings? Also do we really need all of that? It's easy to add something in the public API, but very hard to remove something once it's released. It's ok to "plan" a bit ahead (and keep a few things a bit private even), but we need to be conservative and mindful about the extent of the public API.